hibernate-core/src/main/java/org/hibernate/engine/config/spi/ConfigurationService.java

```diff
@@ -50,7 +52,7 @@ public interface ConfigurationService extends Service {
 	 *
 	 * @return The converted (typed) setting.  Will be the defaultValue if no such setting was defined.
 	 */
-	<T> T getSetting(String name, Converter<T> converter, T defaultValue);
+	<T> @Nullable T getSetting(String name, Converter<T> converter, @Nullable T defaultValue);
```

Comment:
According to https://checkerframework.org/manual/#qualifier-polymorphism we should be able to express conditional nullability with the `@PolyNull` annotation like this:

```suggestion
	<T> @PolyNull T getSetting(String name, Converter<T> converter, @PolyNull T defaultValue);
```

Reply:
It would perhaps not be a terrible idea that you actually explain *why* you would prefer such a construct over the simple `@Nullable`, because now I'm left wondering what the benefit of this would be.

Reply:
By annotating `@PolyNull`, depending on whether you pass a `@Nullable` or `@NonNull` default value, the return type will be `@Nullable` or `@NonNull`, which is exactly what this contract models :)

Reply:
Yes, I understood that, but why, or better said, why *now*, for all I know there might be other similar cases in already nullified code where we did not do this.
Anyway, it's fine.

Reply:
Actually, I have to take that back. Annotating with @PolyNull creates some sort of problem with the Converter, that I was not able to get rid of:
`ConfigurationServiceImpl.java:70: error: [return] incompatible types in return.
                return converter.convert( value );

  type of expression: T extends @Initialized @Nullable Object
  method return type: @org.checkerframework.checker.nullness.qual.PolyNull T extends @Initialized @PolyNull Object
`

Reply:
>why now, for all I know there might be other similar cases in already nullified code where we did not do this.

Because just now I realized that we could do this to avoid unnecessary null checks at use sites.

The error you are seeing is because `Converter#convert` returns something nullable.

I think you need to change the converter to this:

```
    interface Converter<T> {
        /**
         * ...
         */
        @PolyNull T convert(@PolyNull Object value);
    }
```

Reply:
Or alternatively, since this is an SPI, we change that to require a non-nullable return.

```
    interface Converter<T> {
        /**
         * ...
         */
        T convert(@Nullable Object value);
    }
```